akka-http-core/src/main/scala/akka/http/engine/client/HostConnectionPoolSlot.scala

```diff
@@ -0,0 +1,239 @@
+
+      case Cancel     ⇒ { cancel(); shutdown() }
+      case OnComplete ⇒ onComplete()
+      case OnError(e) ⇒ onError(e)
+    }
+
+    def waitingForDemandFromConnection(connInport: ActorRef, connOutport: ActorRef,
+                                       firstRequest: RequestContext): Receive = {
+      case ev @ (Request(_) | Cancel)     ⇒ connOutport ! ev
+      case ev @ (OnComplete | OnError(_)) ⇒ connInport ! ev
+      case OnNext(x)                      ⇒ throw new IllegalStateException("Unrequested RequestContext: " + x)
+
+      case FromConnection(Request(n)) ⇒
+        inflightRequests = inflightRequests.enqueue(firstRequest)
+        request(n - remainingRequested)
+        connInport ! OnNext(firstRequest.request)
+        context.become(running(connInport, connOutport))
+
+      case FromConnection(Cancel)     ⇒ if (!isActive) { cancel(); shutdown() } // else ignore and wait for accompanying OnComplete or OnError
+      case FromConnection(OnComplete) ⇒ handleDisconnect(None)
+      case FromConnection(OnError(e)) ⇒ handleDisconnect(Some(e))
+      case FromConnection(OnNext(x))  ⇒ throw new IllegalStateException("Unexpected HttpResponse: " + x)
+    }
+
+    def running(connInport: ActorRef, connOutport: ActorRef): Receive = {
+      case ev @ (Request(_) | Cancel)     ⇒ connOutport ! ev
+      case ev @ (OnComplete | OnError(_)) ⇒ connInport ! ev
+      case OnNext(rc: RequestContext) ⇒
+        inflightRequests = inflightRequests.enqueue(rc)
+        connInport ! OnNext(rc.request)
+
+      case FromConnection(Request(n)) ⇒ request(n)
+      case FromConnection(Cancel)     ⇒ if (!isActive) { cancel(); shutdown() } // else ignore and wait for accompanying OnComplete or OnError
+
+      case FromConnection(OnNext(response: HttpResponse)) ⇒
+        val requestContext = inflightRequests.head
+        inflightRequests = inflightRequests.tail
+        val delivery = ResponseDelivery(ResponseContext(requestContext, Success(response)))
+        val requestCompleted = SlotEvent.RequestCompleted(slotIx)
+        onNext(delivery :: requestCompleted :: Nil)
+
+      case FromConnection(OnComplete) ⇒ handleDisconnect(None)
+      case FromConnection(OnError(e)) ⇒ handleDisconnect(Some(e))
+    }
+
+    def handleDisconnect(error: Option[Throwable]): Unit = {
+      log.debug("Slot {} disconnected after {}", slotIx, error getOrElse "regular connection close")
+      val results: List[ProcessorOut] = inflightRequests.map { rc ⇒
+        if (rc.retriesLeft == 0) {
+          val reason = error.fold[Throwable](new RuntimeException("Unexpected disconnect"))(identityFunc)
+          ResponseDelivery(ResponseContext(rc, Failure(reason)))
+        } else SlotEvent.RetryRequest(rc.copy(retriesLeft = rc.retriesLeft - 1))
+      }(collection.breakOut)
+      inflightRequests = immutable.Queue.empty
+      onNext(SlotEvent.Disconnected(slotIx, results.size) :: results)
```

Comment:
Right, although I'm not what you mean by not sending empty collections.
I'd say the proper fix is to turn the `SlotEventSplit` into a route that always maintains demand for `SlotEvent`s. The conductor will always demand `SlotEvent`s, so if we repeat the "prepend with doubler-stage" trick here we should be able to always pull out `SlotEvent`s (if no undemanded response is blocking the way, which shouldn't happen).